### PR TITLE
Messaging session unauthorized bug fix

### PR DESCRIPTION
**Pull request by paparekh-amzn**
**Issue #:**
#2446

**Description of changes:**
- Fix behavior of websocket disconnects before a session is connected 
  - Session.start() method promise shall fail if there is a websocket disconnection before a session is established
  - Earlier, session.start() waited till session establish process began, now it blocks till session establish process completes
- Queue messages before SESSION_ESTABLISHED and publish them when a session is established to avoid dropping any messages
  - There was a possibility of dropping messages just in case some message arrived earlier than SESSION_ESTABLISHED message
  - We queue the message now and send it all together when a session is established

**Testing:**
Tested locally by using the messagingSession demo in demos/browser and simulating Unauthorized by providing an invalid UserArn. I saw the expected behavior of promise failing which was not the case before the change.

Tested Queueing of messages and sending it on Session establish by manually initializing a queue with some messages and observed SessionEstablished message along with Queue messages being published on Session establishment together.

Added Unit Tests.

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*

- Setup AWS credentials for an AWS account which has Chime messaging SDK resources
  - vim ~/.aws/credentials and copy access id, secret key and session token
- `cd demos/browser`
- `npm run start --app=messagingSession`
- Go to http://127.0.0.1:8080/ in your Browser and Open Developer Console by Right click -> Inspect -> Console tab
- Provide any invalid UserArn, SessionId is optional and observe console logs using Inspector. There should be a log.error with reason code 4401.

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
Yes

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
No

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
Yes

### CODE REVIEW

File: src/messagingsession/DefaultMessagingSession.ts [hunk 1]
```diff
@@ -107,6 +111,14 @@ export default class DefaultMessagingSession implements MessagingSession {
   }
 
   private async startConnecting(reconnecting: boolean): Promise<void> {
+    await this.startConnectingInternal(reconnecting);
+    return await new Promise<void>((resolve, reject) => {
+      this.bootstrapResolved = resolve;
```
Comment (dpwspoon) [inline src/messagingsession/DefaultMessagingSession.ts:116]: Think you can just store the promise directly and then use .resolve or .reject on it: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject

Comment (paparekh-amzn) [reply src/messagingsession/DefaultMessagingSession.ts:116]: Tried with the approach you mentioned and faced some issues before I went with this approach recommended in SO.

File: src/messagingsession/DefaultMessagingSession.ts [hunk 2]
```diff
@@ -224,24 +236,39 @@ export default class DefaultMessagingSession implements MessagingSession {
...
+          });
+        }
       } else if (!this.isSessionEstablished) {
         // SESSION_ESTABLISHED is not guaranteed to be the first message, and in rare conditions a message or event from
         // a channel the member is a member of might arrive prior to SESSION_ESTABLISHED.  Because SESSION_ESTABLISHED indicates
-        // it is safe to bootstrap the user application with out any race conditions in losing events we opt to drop messages prior
-        // to SESSION_ESTABLISHED being received
+        // it is safe to bootstrap the user application without any race conditions in losing events we opt to store messages prior
+        // to SESSION_ESTABLISHED being received and send when once SESSION_ESTABLISHED.
+        this.preBootstrapMessages.push(message);
```
Comment (ltrung) [inline src/messagingsession/DefaultMessagingSession.ts:255]: On average  how many messages would be in this queue and how often this happens? Do we need to limit the size of this queue?

File: demos/browser/package-lock.json [hunk 3]
```diff
@@ -1,7204 +1,8 @@
 {
   "name": "amazon-chime-sdk-js-demo",
   "version": "2.0.0",
-  "lockfileVersion": 2,
```
Comment (simmkyu) [inline demos/browser/package-lock.json:4]: Please use the later versions of NPM. The`lockfileVersion` should remain the same.

File: src/messagingsession/DefaultMessagingSession.ts (hunk 1 above)
Comment (simmkyu) [inline src/messagingsession/DefaultMessagingSession.ts:116]: In a slow (or almost offline) internet connection, the `startConnecting` will take a while to resolve. I wonder if we need to add a timeout or retry logic.

Also, if customers rely on the existing behavior, they may experience unexpected user experiences. For example, if customers update the UI after `await messagingSession.start()`, they will see some delay now.

Comment (paparekh-amzn) [reply src/messagingsession/DefaultMessagingSession.ts:116]: This seems like an existing issue even without this change. I agree that a timeout or retry logic shall be helpful but I think that could be tackled separately considering this fix is a result of a customer ticket.

File: demos/browser/package-lock.json (hunk 3 above)
Comment (paparekh-amzn) [reply demos/browser/package-lock.json:4]: Sure, will do

File: src/messagingsession/DefaultMessagingSession.ts (hunk 2 above)
Comment (paparekh-amzn) [reply src/messagingsession/DefaultMessagingSession.ts:255]: We believe this to be a rare race condition after introducing prefetch feature. Maximum number of messages we sent via Prefetch is 50.

File: src/messagingsession/DefaultMessagingSession.ts (hunk 1 above)
Comment (simmkyu) [reply src/messagingsession/DefaultMessagingSession.ts:116]: Right. The `startConnecting` previously called other async APIs like `getMessagingSessionEndpoint`. `getMessagingSessionEndpoint` will be slow too.

**Review by ltrung — COMMENTED** [review]
Can we update the testing instructions step-by-step so QA could follow?

**Review by ltrung — APPROVED** [review]

File: src/messagingsession/DefaultMessagingSession.ts [hunk 4]
```diff
@@ -51,6 +54,7 @@ export default class DefaultMessagingSession implements MessagingSession {
 
     CSPMonitor.addLogger(this.logger);
     CSPMonitor.register();
+    this.preBootstrapMessages = [];
```
Comment (dpwspoon) [inline src/messagingsession/DefaultMessagingSession.ts:57]: nit bootstrap isn't a thing.  Let's use a different term like sessionEstablished

**Review by simmkyu — APPROVED** [review]

**Review by dpwspoon — APPROVED** [review]

**Pull request merged**